In [1]:
import pandas as pd
import numpy as np
import requests
import zipfile
from pathlib import Path
import pandas as pd
from io import BytesIO, StringIO
import time
import os
from datetime import datetime, date, time 

In [2]:
folder_path = "D:\\Jai - NSS\\bna-airline-delay-patterns\\bna-delay-cause"

# List to store all DataFrames
all_dfs = []

for file_name in os.listdir(folder_path):
    if file_name.endswith(".zip"):
        zip_path = os.path.join(folder_path, file_name)
        
        with zipfile.ZipFile(zip_path, 'r') as zip_file:
            for inner_file in zip_file.namelist():
                # Check if the inner file is a CSV
                if inner_file.endswith(".csv"):
                    with zip_file.open(inner_file) as f:
                        # Read bytes and decode to string
                        data = f.read().decode('utf-8')
                        
                        # Load CSV into a DataFrame
                        df = pd.read_csv(StringIO(data))
                        all_dfs.append(df)

# Combine all DataFrames into one
combined_df = pd.concat(all_dfs, ignore_index=True)
print(combined_df.head())
#https://www.geeksforgeeks.org/python/working-zip-files-python/
##little side note 'utf-8' is how python reads bytes to string. If you do not add 'utf-8' it will read all the data raw. There are other like 'latin-1', 'iso-8859-1' ,'utf-16' If you try to decode a file with the wrong encoding, Python will raise a UnicodeDecodeError

   YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK               FL_DATE  \
0  2024        1      1             1            1  1/1/2024 12:00:00 AM   
1  2024        1      1             1            1  1/1/2024 12:00:00 AM   
2  2024        1      1             1            1  1/1/2024 12:00:00 AM   
3  2024        1      1             1            1  1/1/2024 12:00:00 AM   
4  2024        1      1             1            1  1/1/2024 12:00:00 AM   

  BRANDED_CODE_SHARE ORIGIN DEST  CRS_DEP_TIME  ...  CANCELLED  \
0                 AA    BNA  CLT           537  ...        0.0   
1                 AA    BNA  CLT           732  ...        0.0   
2                 AA    BNA  CLT           943  ...        0.0   
3                 AA    BNA  CLT          1927  ...        0.0   
4                 AA    BNA  DFW           501  ...        0.0   

   CANCELLATION_CODE DIVERTED  FLIGHTS  DISTANCE  CARRIER_DELAY WEATHER_DELAY  \
0                NaN      0.0      1.0     328.0            NaN  

In [3]:
combined_df

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK  \
0       2024        1      1             1            1   
1       2024        1      1             1            1   
2       2024        1      1             1            1   
3       2024        1      1             1            1   
4       2024        1      1             1            1   
...      ...      ...    ...           ...          ...   
887288  2022        4     12            31            6   
887289  2022        4     12            31            6   
887290  2022        4     12            31            6   
887291  2022        4     12            31            6   
887292  2022        4     12            31            6   

                       FL_DATE BRANDED_CODE_SHARE ORIGIN DEST  CRS_DEP_TIME  \
0         1/1/2024 12:00:00 AM                 AA    BNA  CLT           537   
1         1/1/2024 12:00:00 AM                 AA    BNA  CLT           732   
2         1/1/2024 12:00:00 AM                 AA    BNA  CLT           943   
3         1/1/2024 12:00:00 AM                 AA    BNA  CLT          1927   
4         1/1/2024 12:00:00 AM                 AA    BNA  DFW           501   
...                        ...                ...    ...  ...           ...   
887288  12/31/2022 12:00:00 AM                 WN    TPA  BNA          1030   
887289  12/31/2022 12:00:00 AM                 WN    TPA  BNA          1255   
887290  12/31/2022 12:00:00 AM                 WN    TPA  BNA          1750   
887291  12/31/2022 12:00:00 AM                 WN    TPA  BNA          1910   
887292  12/31/2022 12:00:00 AM                 WN    TPA  MEM          1550   

        ...  CANCELLED  CANCELLATION_CODE DIVERTED  FLIGHTS  DISTANCE  \
0       ...        0.0                NaN      0.0      1.0     328.0   
1       ...        0.0                NaN      0.0      1.0     328.0   
2       ...        0.0                NaN      0.0      1.0     328.0   
3       ...        0.0                NaN      0.0      1.0     328.0   
4       ...        0.0                NaN      0.0      1.0     631.0   
...     ...        ...                ...      ...      ...       ...   
887288  ...        0.0                NaN      0.0      1.0     612.0   
887289  ...        0.0                NaN      0.0      1.0     612.0   
887290  ...        0.0                NaN      0.0      1.0     612.0   
887291  ...        0.0                NaN      0.0      1.0     612.0   
887292  ...        0.0                NaN      0.0      1.0     655.0   

        CARRIER_DELAY WEATHER_DELAY  NAS_DELAY SECURITY_DELAY  \
0                 NaN           NaN        NaN            NaN   
1                 NaN           NaN        NaN            NaN   
2                 NaN           NaN        NaN            NaN   
3                 NaN           NaN        NaN            NaN   
4                 NaN           NaN        NaN            NaN   
...               ...           ...        ...            ...   
887288            NaN           NaN        NaN            NaN   
887289            NaN           NaN        NaN            NaN   
887290            NaN           NaN        NaN            NaN   
887291            4.0           0.0        0.0            0.0   
887292           29.0           0.0        0.0            0.0   

        LATE_AIRCRAFT_DELAY  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  
...                     ...  
887288                  NaN  
887289                  NaN  
887290                  NaN  
887291                 19.0  
887292                 77.0  

[887293 rows x 27 columns]

In [4]:
combined_df.columns = combined_df.columns.str.lower().str.strip()

In [5]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887293 entries, 0 to 887292
Data columns (total 27 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   year                 887293 non-null  int64  
 1   quarter              887293 non-null  int64  
 2   month                887293 non-null  int64  
 3   day_of_month         887293 non-null  int64  
 4   day_of_week          887293 non-null  int64  
 5   fl_date              887293 non-null  object 
 6   branded_code_share   887293 non-null  object 
 7   origin               887293 non-null  object 
 8   dest                 887293 non-null  object 
 9   crs_dep_time         887293 non-null  int64  
 10  dep_delay_new        872711 non-null  float64
 11  dep_del15            872711 non-null  float64
 12  dep_time_blk         887293 non-null  object 
 13  crs_arr_time         887293 non-null  int64  
 14  arr_delay_new        870506 non-null  float64
 15  arr_del15        

In [79]:
combined_df.fl_date.nunique() 

1095

In [78]:
combined_df['fl_date'].info()

<class 'pandas.core.series.Series'>
Index: 886483 entries, 0 to 887292
Series name: fl_date
Non-Null Count   Dtype 
--------------   ----- 
886483 non-null  object
dtypes: object(1)
memory usage: 13.5+ MB


#### 3*365 = 1095 days. There is an extra day that needs to be looked at
##### There is a leap year in 02/29/2024. Like the niche airlines let remove the niche day. It only happens every 4 years.

In [9]:
combined_df['fl_date'] = pd.to_datetime(combined_df['fl_date'], errors='coerce')
combined_df = combined_df[combined_df['fl_date'] != '2024-02-29']
combined_df['fl_date'] = combined_df['fl_date'].dt.date

C:\Users\Cajun\AppData\Local\Temp\ipykernel_17696\2202483181.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  combined_df['fl_date'] = pd.to_datetime(combined_df['fl_date'], errors='coerce')
C:\Users\Cajun\AppData\Local\Temp\ipykernel_17696\2202483181.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  combined_df['fl_date'] = combined_df['fl_date'].dt.date


In [10]:
bna_df = combined_df[(combined_df['origin'] == 'BNA')]
bna_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 285863 entries, 0 to 887175
Data columns (total 27 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   year                 285863 non-null  int64  
 1   quarter              285863 non-null  int64  
 2   month                285863 non-null  int64  
 3   day_of_month         285863 non-null  int64  
 4   day_of_week          285863 non-null  int64  
 5   fl_date              285863 non-null  object 
 6   branded_code_share   285863 non-null  object 
 7   origin               285863 non-null  object 
 8   dest                 285863 non-null  object 
 9   crs_dep_time         285863 non-null  int64  
 10  dep_delay_new        280768 non-null  float64
 11  dep_del15            280768 non-null  float64
 12  dep_time_blk         285863 non-null  object 
 13  crs_arr_time         285863 non-null  int64  
 14  arr_delay_new        280046 non-null  float64
 15  arr_del15            2

In [84]:
bna_df = bna_df.drop_duplicates()
print(f"Formerly there were 1096 days. Now there are {bna_df['fl_date'].nunique()} days")

Formerly there were 1096 days. Now there are 1095 days


In [82]:
missing_data = bna_df[bna_df['month'] == 2]
missing_data.fl_date.nunique() 
print(f'Formerly there were 85 days. Now there are {missing_data.fl_date.nunique()} days')

Formerly there were 85 days. Now there are 84 days


### Need to confirm if the date ranges are missing one day

In [13]:
full_range = pd.date_range(start = bna_df['fl_date'].min(), end = bna_df['fl_date'].max())
full_range
## https://pandas.pydata.org/docs/reference/api/pandas.date_range.html

DatetimeIndex(['2022-01-01', '2022-01-02', '2022-01-03', '2022-01-04',
               '2022-01-05', '2022-01-06', '2022-01-07', '2022-01-08',
               '2022-01-09', '2022-01-10',
               ...
               '2024-12-22', '2024-12-23', '2024-12-24', '2024-12-25',
               '2024-12-26', '2024-12-27', '2024-12-28', '2024-12-29',
               '2024-12-30', '2024-12-31'],
              dtype='datetime64[ns]', length=1096, freq='D')

In [14]:
existing_dates = pd.to_datetime(bna_df['fl_date'].unique())
existing_dates

DatetimeIndex(['2024-01-01', '2024-01-02', '2024-01-03', '2024-01-04',
               '2024-01-05', '2024-01-06', '2024-01-07', '2024-01-08',
               '2024-01-09', '2024-01-10',
               ...
               '2022-12-22', '2022-12-23', '2022-12-24', '2022-12-25',
               '2022-12-26', '2022-12-27', '2022-12-28', '2022-12-29',
               '2022-12-30', '2022-12-31'],
              dtype='datetime64[ns]', length=1095, freq=None)

In [15]:
missing_dates = full_range.difference(existing_dates) ##The .difference() method returns a set that contains the difference between two sets.
print(f"Total missing days: {len(missing_dates)}") ## We have one day missing. Which is 2/29/2024

Total missing days: 1


In [16]:
bna_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 285863 entries, 0 to 887175
Data columns (total 27 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   year                 285863 non-null  int64  
 1   quarter              285863 non-null  int64  
 2   month                285863 non-null  int64  
 3   day_of_month         285863 non-null  int64  
 4   day_of_week          285863 non-null  int64  
 5   fl_date              285863 non-null  object 
 6   branded_code_share   285863 non-null  object 
 7   origin               285863 non-null  object 
 8   dest                 285863 non-null  object 
 9   crs_dep_time         285863 non-null  int64  
 10  dep_delay_new        280768 non-null  float64
 11  dep_del15            280768 non-null  float64
 12  dep_time_blk         285863 non-null  object 
 13  crs_arr_time         285863 non-null  int64  
 14  arr_delay_new        280046 non-null  float64
 15  arr_del15            2

### Let's look at the airlines that share a code

In [17]:
bna_df['branded_code_share'].unique() 

array(['AA', 'AA_CODESHARE', 'AS', 'B6', 'DL', 'DL_CODESHARE', 'F9', 'G4',
       'NK', 'UA', 'UA_CODESHARE', 'WN'], dtype=object)

##### Codeshare flights are operated by smaller regional airlines but sold under the brand of a larger airline. 
##### When looking at the data 

In [18]:
bna_df = bna_df[~bna_df['branded_code_share'].str.contains('_CODESHARE')]

In [19]:
bna_df['branded_code_share'].unique()

array(['AA', 'AS', 'B6', 'DL', 'F9', 'G4', 'NK', 'UA', 'WN'], dtype=object)

#### There are two common ways to assign seasons to months:
     1. Using a dictionary to map months to seasons
     2. Creating a user-defined function
#### Here’s an example using a dictionary:
season_map = {1: "Winter", 2: "Winter",3: "Spring", 4: "Spring", 5: "Spring", 6: "Summer", 7: "Summer", 8: "Summer", 9: "Fall", 10: "Fall", 11: "Fall", 12: "Winter"}
##### The map() function can then be used to apply this mapping to a pandas Series. It applies a given function or mapping to each element of an iterable (like a list or Series) and returns a new mapped object.

##### For more details, see: https://www.geeksforgeeks.org/python/python-map-function/

In [51]:
def season_map(season):
    if season <=2:
        return 'winter'
    elif 3 <= season <=5:
        return 'spring'
    elif 6 <= season <= 8:
        return 'summer'
    elif 9 <= season <= 11:
        return 'fall'
    else:
        return 'winter'


bna_df['season'] = bna_df['month'].map(season_map)
bna_df.sample(5)

year  quarter  month  day_of_month  day_of_week     fl_date  \
499622  2023        3      8            17            4  2023-08-17   
114099  2024        2      5            21            2  2024-05-21   
110750  2024        2      5            17            5  2024-05-17   
683047  2022        2      4             9            6  2022-04-09   
211828  2024        3      9             5            4  2024-09-05   

       branded_code_share origin dest  crs_dep_time  ...  distance  \
499622                 WN    BNA  DAL          1825  ...     623.0   
114099                 UA    BNA  ORD          1255  ...     409.0   
110750                 WN    BNA  IAH          1245  ...     657.0   
683047                 WN    BNA  DAL          1515  ...     623.0   
211828                 AA    BNA  CLT           700  ...     328.0   

        carrier_delay weather_delay  nas_delay  security_delay  \
499622           21.0           0.0        0.0             0.0   
114099            NaN           NaN        NaN             NaN   
110750            0.0           0.0        4.0             0.0   
683047            NaN           NaN        NaN             NaN   
211828            NaN           NaN        NaN             NaN   

        late_aircraft_delay  season      time_of_day       day  week_type  
499622                 15.0  summer    Early Evening  Thursday    weekday  
114099                  NaN  spring  Early Afternoon   Tuesday    weekday  
110750                 17.0  spring  Early Afternoon    Friday    weekday  
683047                  NaN  spring        Afternoon  Saturday    weekend  
211828                  NaN    fall          Morning  Thursday    weekday  

[5 rows x 31 columns]

In [52]:
bna_df['crs_arr_time'].unique()

array([ 8,  9, 12, 21,  7, 10, 13, 15, 17, 19, 11, 20, 16, 14, 22,  0, 23,
       18,  1,  6])

In [53]:
##Morning: Generally considered to start around 6 AM and end around 12 PM (noon).
##Early Morning: 5 AM to 8 AM.
##Late Morning: 11 AM to 12 PM.
##Afternoon: 12 PM (noon) to 6 PM.
##Early Afternoon: 1 PM to 3 PM.
##Late Afternoon: 4 PM to 5 PM.
##Evening: 6 PM to 12 AM (midnight).
##Night: 12 AM (midnight) to 6 AM.


time_map = {0: 'Night',
            1: 'Night',
            2: 'Night',
            3: 'Night',
            4: 'Night',
            5: 'Early Morning',
            6: 'Early Morning',
            7: 'Early Morning',
            8: 'Early Morning',
            9: 'Morning',
            10:'Morning',
            11:'Morning',
            12:'Early Afternoon',
            13:'Early Afternoon',
            14:'Early Afternoon',
            15:'Afternoon',
            16:'Afternoon',
            17:'Afternoon',
            19:'Early Evening',
            20:'Early Evening',
            21:'Evening',
            22:'Evening',
            23:'Evening'}


bna_df['crs_arr_time'].tolist()
bna_df['crs_arr_time'] = bna_df['crs_arr_time'].fillna(0).astype(int)
bna_df['crs_arr_time'] = bna_df['crs_arr_time']//100
bna_df['time_of_day'] = bna_df['crs_arr_time'].map(time_map)
bna_df


year  quarter  month  day_of_month  day_of_week     fl_date  \
0       2024        1      1             1            1  2024-01-01   
1       2024        1      1             1            1  2024-01-01   
2       2024        1      1             1            1  2024-01-01   
3       2024        1      1             1            1  2024-01-01   
4       2024        1      1             1            1  2024-01-01   
...      ...      ...    ...           ...          ...         ...   
887171  2022        4     12            31            6  2022-12-31   
887172  2022        4     12            31            6  2022-12-31   
887173  2022        4     12            31            6  2022-12-31   
887174  2022        4     12            31            6  2022-12-31   
887175  2022        4     12            31            6  2022-12-31   

       branded_code_share origin dest  crs_dep_time  ...  distance  \
0                      AA    BNA  CLT           537  ...     328.0   
1                      AA    BNA  CLT           732  ...     328.0   
2                      AA    BNA  CLT           943  ...     328.0   
3                      AA    BNA  CLT          1927  ...     328.0   
4                      AA    BNA  DFW           501  ...     631.0   
...                   ...    ...  ...           ...  ...       ...   
887171                 WN    BNA  TPA           500  ...     612.0   
887172                 WN    BNA  TPA           850  ...     612.0   
887173                 WN    BNA  TPA          1455  ...     612.0   
887174                 WN    BNA  TPA          1735  ...     612.0   
887175                 WN    BNA  TPA          2045  ...     612.0   

        carrier_delay weather_delay  nas_delay  security_delay  \
0                 NaN           NaN        NaN             NaN   
1                 NaN           NaN        NaN             NaN   
2                 NaN           NaN        NaN             NaN   
3                 NaN           NaN        NaN             NaN   
4                 NaN           NaN        NaN             NaN   
...               ...           ...        ...             ...   
887171            NaN           NaN        NaN             NaN   
887172            0.0           0.0       18.0             0.0   
887173            NaN           NaN        NaN             NaN   
887174            NaN           NaN        NaN             NaN   
887175            NaN           NaN        NaN             NaN   

        late_aircraft_delay  season  time_of_day       day  week_type  
0                       NaN  winter        Night    Monday    weekday  
1                       NaN  winter        Night    Monday    weekday  
2                       NaN  winter        Night    Monday    weekday  
3                       NaN  winter        Night    Monday    weekday  
4                       NaN  winter        Night    Monday    weekday  
...                     ...     ...          ...       ...        ...  
887171                  NaN  winter        Night  Saturday    weekend  
887172                  0.0  winter        Night  Saturday    weekend  
887173                  NaN  winter        Night  Saturday    weekend  
887174                  NaN  winter        Night  Saturday    weekend  
887175                  NaN  winter        Night  Saturday    weekend  

[237029 rows x 31 columns]

Code	Description
#2	Tuesday
#3	Wednesday
#4	Thursday
#5	Friday
#6	Saturday
#7	Sunday
#From the bureau of transportation

In [54]:
day_map = {1:'Monday',
           2:'Tuesday',
           3:'Wednesday',
           4:'Thursday',
           5:'Friday',
           6:'Saturday',
           7:'Sunday'}

bna_df['day'] = bna_df['day_of_week'].map(day_map)
bna_df.sample(5)


year  quarter  month  day_of_month  day_of_week     fl_date  \
264355  2024        4     11             1            5  2024-11-01   
584337  2023        4     11            24            5  2023-11-24   
383526  2023        1      3            30            4  2023-03-30   
575299  2023        4     11            13            1  2023-11-13   
600091  2023        4     12            13            3  2023-12-13   

       branded_code_share origin dest  crs_dep_time  ...  distance  \
264355                 WN    BNA  GSP          1720  ...     265.0   
584337                 WN    BNA  JAX          1945  ...     483.0   
383526                 WN    BNA  MIA          2120  ...     806.0   
575299                 WN    BNA  LAX          1900  ...    1797.0   
600091                 WN    BNA  FLL           640  ...     793.0   

        carrier_delay weather_delay  nas_delay  security_delay  \
264355            NaN           NaN        NaN             NaN   
584337            NaN           NaN        NaN             NaN   
383526            NaN           NaN        NaN             NaN   
575299            NaN           NaN        NaN             NaN   
600091            0.0           0.0       16.0             0.0   

        late_aircraft_delay  season  time_of_day        day  week_type  
264355                  NaN    fall        Night     Friday    weekday  
584337                  NaN    fall        Night     Friday    weekday  
383526                  NaN  spring        Night   Thursday    weekday  
575299                  NaN    fall        Night     Monday    weekday  
600091                  0.0  winter        Night  Wednesday    weekday  

[5 rows x 31 columns]

#### labeling the weekend or weekend

In [56]:
def week_type(day):
    if day <= 5:
        return 'weekday'
    else:
        return 'weekend'
        
bna_df['week_type'] = bna_df['day_of_week'].map(week_type)
bna_df.sample(5)

year  quarter  month  day_of_month  day_of_week     fl_date  \
191698  2024        3      8            12            1  2024-08-12   
530638  2023        3      9            24            7  2023-09-24   
9625    2024        1      1            13            6  2024-01-13   
250885  2024        4     10            18            5  2024-10-18   
786683  2022        3      8            22            1  2022-08-22   

       branded_code_share origin dest  crs_dep_time  ...  distance  \
191698                 WN    BNA  DEN           750  ...    1014.0   
530638                 AA    BNA  LGA           745  ...     764.0   
9625                   WN    BNA  CHS           955  ...     439.0   
250885                 AA    BNA  MIA          1829  ...     806.0   
786683                 WN    BNA  MSY          1310  ...     471.0   

        carrier_delay weather_delay  nas_delay  security_delay  \
191698            NaN           NaN        NaN             NaN   
530638           43.0           0.0        0.0             0.0   
9625              NaN           NaN        NaN             NaN   
250885            NaN           NaN        NaN             NaN   
786683            NaN           NaN        NaN             NaN   

        late_aircraft_delay  season  time_of_day       day  week_type  
191698                  NaN  summer        Night    Monday    weekday  
530638                  0.0    fall        Night    Sunday    weekend  
9625                    NaN  winter        Night  Saturday    weekend  
250885                  NaN    fall        Night    Friday    weekday  
786683                  NaN  summer        Night    Monday    weekday  

[5 rows x 31 columns]

#### Lets look at the null values

In [57]:
percent_missing = bna_df.isnull().sum() * 100 / len(bna_df)
missing_value_df = pd.DataFrame({'column_name': bna_df.columns,
                                 'percent_missing': percent_missing})
missing_value_df

column_name  percent_missing
year                                year         0.000000
quarter                          quarter         0.000000
month                              month         0.000000
day_of_month                day_of_month         0.000000
day_of_week                  day_of_week         0.000000
fl_date                          fl_date         0.000000
branded_code_share    branded_code_share         0.000000
origin                            origin         0.000000
dest                                dest         0.000000
crs_dep_time                crs_dep_time         0.000000
dep_delay_new              dep_delay_new         1.697683
dep_del15                      dep_del15         1.697683
dep_time_blk                dep_time_blk         0.000000
crs_arr_time                crs_arr_time         0.000000
arr_delay_new              arr_delay_new         1.929722
arr_del15                      arr_del15         1.929722
arr_time_blk                arr_time_blk         0.000000
cancelled                      cancelled         0.000000
cancellation_code      cancellation_code        98.276160
diverted                        diverted         0.000000
flights                          flights         0.000000
distance                        distance         0.000000
carrier_delay              carrier_delay        80.299457
weather_delay              weather_delay        80.299457
nas_delay                      nas_delay        80.299457
security_delay            security_delay        80.299457
late_aircraft_delay  late_aircraft_delay        80.299457
season                            season         0.000000
time_of_day                  time_of_day         0.000000
day                                  day         0.000000
week_type                      week_type         0.000000

# THIS IS  ON TIME PERFORMANCE

In [58]:
carrier_performance_agg = bna_df.groupby('branded_code_share').agg(totalflights = ('flights', 'sum'),
                                                             arr_delayedflights = ('arr_del15','sum'),
                                                             arr_delayedmin = ('arr_delay_new','sum'),
                                                             dep_delayedmin = ('dep_delay_new','sum'),
                                                             dep_delayedflights = ('dep_del15','sum'))     
carrier_performance_agg     

totalflights  arr_delayedflights  arr_delayedmin  \
branded_code_share                                                     
AA                       23785.0              5349.0        523529.0   
AS                        2740.0               551.0         31201.0   
B6                        5768.0              1779.0        161878.0   
DL                       24010.0              2776.0        235936.0   
F9                        2779.0               809.0         58565.0   
G4                        9560.0              2274.0        184703.0   
NK                        9507.0              2574.0        200013.0   
UA                       16215.0              3227.0        277432.0   
WN                      142665.0             27357.0       1561098.0   

                    dep_delayedmin  dep_delayedflights  
branded_code_share                                      
AA                        510098.0              4976.0  
AS                         33140.0               590.0  
B6                        164054.0              1786.0  
DL                        241573.0              2781.0  
F9                         56066.0               758.0  
G4                        170835.0              2056.0  
NK                        203828.0              2600.0  
UA                        268672.0              3018.0  
WN                       1819223.0             32504.0

In [59]:
carrier_performance_agg['Arrival Flight Delays (%)'] = ((carrier_performance_agg['arr_delayedflights']/carrier_performance_agg['totalflights'])*100).round(2)
carrier_performance_agg['Departure Flight Delays(%)'] = ((carrier_performance_agg['dep_delayedflights']/carrier_performance_agg['totalflights'])*100).round(2)
carrier_performance_agg['Avg. arrival delay (min)'] = (carrier_performance_agg['arr_delayedmin']/carrier_performance_agg['arr_delayedflights'])
carrier_performance_agg['Avg. departure delay(min)'] = (carrier_performance_agg['dep_delayedmin']/carrier_performance_agg['dep_delayedflights'])
carrier_performance_agg = carrier_performance_agg.sort_values(by = ['Arrival Flight Delays (%)','Departure Flight Delays(%)'], ascending = [False, False])

In [60]:
carrier_performance_agg = carrier_performance_agg.rename({'totalflights': '# of total flights', 
                                                          'arr_delayedflights': '# of total arrival flights',
                                                          'arr_delayedmin': '# total arrival delayed min',
                                                          'dep_delayedmin': '# total departure delayed min',
                                                          'dep_delayedflights': '# of total departure flights'})
carrier_performance_agg 

totalflights  arr_delayedflights  arr_delayedmin  \
branded_code_share                                                     
B6                        5768.0              1779.0        161878.0   
F9                        2779.0               809.0         58565.0   
NK                        9507.0              2574.0        200013.0   
G4                        9560.0              2274.0        184703.0   
AA                       23785.0              5349.0        523529.0   
AS                        2740.0               551.0         31201.0   
UA                       16215.0              3227.0        277432.0   
WN                      142665.0             27357.0       1561098.0   
DL                       24010.0              2776.0        235936.0   

                    dep_delayedmin  dep_delayedflights  \
branded_code_share                                       
B6                        164054.0              1786.0   
F9                         56066.0               758.0   
NK                        203828.0              2600.0   
G4                        170835.0              2056.0   
AA                        510098.0              4976.0   
AS                         33140.0               590.0   
UA                        268672.0              3018.0   
WN                       1819223.0             32504.0   
DL                        241573.0              2781.0   

                    Arrival Flight Delays (%)  Departure Flight Delays(%)  \
branded_code_share                                                          
B6                                      30.84                       30.96   
F9                                      29.11                       27.28   
NK                                      27.07                       27.35   
G4                                      23.79                       21.51   
AA                                      22.49                       20.92   
AS                                      20.11                       21.53   
UA                                      19.90                       18.61   
WN                                      19.18                       22.78   
DL                                      11.56                       11.58   

                    Avg. arrival delay (min)  Avg. departure delay(min)  
branded_code_share                                                       
B6                                 90.993817                  91.855543  
F9                                 72.391842                  73.965699  
NK                                 77.705128                  78.395385  
G4                                 81.223835                  83.090953  
AA                                 97.874182                 102.511656  
AS                                 56.626134                  56.169492  
UA                                 85.972110                  89.023194  
WN                                 57.063932                  55.969204  
DL                                 84.991354                  86.865516

# Overall delay rate

In [61]:
avg_arr_delay_grouped = bna_df.groupby('branded_code_share')['arr_delay_new'].mean().reset_index()
avg_dep_delay_grouped = bna_df.groupby('branded_code_share')['dep_delay_new'].mean().reset_index()
avg_delay_grouped = pd.merge(avg_arr_delay_grouped, avg_dep_delay_grouped, on = 'branded_code_share')


In [62]:
delay_summary = (
    bna_df.groupby(['branded_code_share', 'season', 'time_of_day'])
    .agg(total_flights=('flights', 'sum'),
        delayed_arrivals=('arr_del15', 'sum'),
        delayed_departures=('dep_del15', 'sum'))
    .reset_index())


delay_summary['arrival_delay_rate (%)'] = (delay_summary['delayed_arrivals'] / delay_summary['total_flights'] * 100).round(2)
delay_summary['departure_delay_rate (%)'] = (delay_summary['delayed_departures'] / delay_summary['total_flights'] * 100).round(2)
delay_summary['overall_delay_rate (%)'] = ((delay_summary['delayed_arrivals'] + delay_summary['delayed_departures']) / (2*delay_summary['total_flights']) * 100).round(2)


# monthly_df

In [68]:
# df1: Monthly percentages of delayed flights
df1 = (bna_df.groupby(['branded_code_share'])
       [['arr_del15', 'dep_del15', 'flights']]
       .sum()
       .reset_index())
df1['arrival flights monthly (%)'] = (df1['arr_del15'] / df1['flights']).round(4) * 100
df1['departure flights monthly (%)'] = (df1['dep_del15'] / df1['flights']).round(4) * 100
df1 = df1[['branded_code_share','arrival flights monthly (%)','departure flights monthly (%)']]

# df2: Yearly percentages of delayed flights
df2 = (bna_df.groupby(['branded_code_share'])
       [['arr_del15','dep_del15','flights']]
       .sum()
       .reset_index())
df2['arrival flights yearly (%)'] = (df2['arr_del15'] / df2['flights']).round(4) * 100
df2['departure flights yearly (%)'] = (df2['dep_del15'] / df2['flights']).round(4) * 100
df2 = df2[['branded_code_share','arrival flights yearly (%)','departure flights yearly (%)']]

# df3: Yearly average minutes delayed
df3 = (bna_df.groupby(['branded_code_share'])
       [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
       .sum()
       .reset_index())
df3['Avg. min delayed for arrival flights (year)'] = df3['arr_delay_new'] / df3['arr_del15']
df3['Avg. min delayed for departure flights (year)'] = df3['dep_delay_new'] / df3['dep_del15']
df3 = df3[['branded_code_share','Avg. min delayed for arrival flights (year)','Avg. min delayed for departure flights (year)']]

# df4: Monthly average minutes delayed
df4 = (bna_df.groupby(['branded_code_share'])
       [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
       .sum()
       .reset_index())
df4['Avg. min delayed for arrival flights (month)'] = df4['arr_delay_new'] / df4['arr_del15']
df4['Avg. min delayed for departure flights (month)'] = df4['dep_delay_new'] / df4['dep_del15']
df4 = df4[['branded_code_share','Avg. min delayed for arrival flights (month)','Avg. min delayed for departure flights (month)']]

# Merge all summaries into a single DataFrame
week_df = df1.merge(df2, on='branded_code_share', how='outer') \
             .merge(df3, on='branded_code_share', how='outer') \
             .merge(df4, on='branded_code_share', how='outer')

In [69]:
week_df

branded_code_share  arrival flights monthly (%)  \
0                 AA                        22.49   
1                 AS                        20.11   
2                 B6                        30.84   
3                 DL                        11.56   
4                 F9                        29.11   
5                 G4                        23.79   
6                 NK                        27.07   
7                 UA                        19.90   
8                 WN                        19.18   

   departure flights monthly (%)  arrival flights yearly (%)  \
0                          20.92                       22.49   
1                          21.53                       20.11   
2                          30.96                       30.84   
3                          11.58                       11.56   
4                          27.28                       29.11   
5                          21.51                       23.79   
6                          27.35                       27.07   
7                          18.61                       19.90   
8                          22.78                       19.18   

   departure flights yearly (%)  Avg. min delayed for arrival flights (year)  \
0                         20.92                                    97.874182   
1                         21.53                                    56.626134   
2                         30.96                                    90.993817   
3                         11.58                                    84.991354   
4                         27.28                                    72.391842   
5                         21.51                                    81.223835   
6                         27.35                                    77.705128   
7                         18.61                                    85.972110   
8                         22.78                                    57.063932   

   Avg. min delayed for departure flights (year)  \
0                                     102.511656   
1                                      56.169492   
2                                      91.855543   
3                                      86.865516   
4                                      73.965699   
5                                      83.090953   
6                                      78.395385   
7                                      89.023194   
8                                      55.969204   

   Avg. min delayed for arrival flights (month)  \
0                                     97.874182   
1                                     56.626134   
2                                     90.993817   
3                                     84.991354   
4                                     72.391842   
5                                     81.223835   
6                                     77.705128   
7                                     85.972110   
8                                     57.063932   

   Avg. min delayed for departure flights (month)  
0                                      102.511656  
1                                       56.169492  
2                                       91.855543  
3                                       86.865516  
4                                       73.965699  
5                                       83.090953  
6                                       78.395385  
7                                       89.023194  
8                                       55.969204

# Season DF

In [72]:
season_df = (bna_df.groupby(['branded_code_share', 'season', 'year', 'month'])
             [['arr_delay_new', 'dep_delay_new', 'arr_del15', 'dep_del15', 'flights']]
             .sum()
             .reset_index())

# Average minutes delayed
season_df['Avg. min delayed for arrival flights (season)'] = season_df['arr_delay_new'] / season_df['arr_del15']
season_df['Avg. min delayed for departure flights (season)'] = season_df['dep_delay_new'] / season_df['dep_del15']

# Percent of flights delayed
season_df['arrival flights season (%)'] = (season_df['arr_del15'] / season_df['flights']).round(4) * 100
season_df['departure flights season (%)'] = (season_df['dep_del15'] / season_df['flights']).round(4) * 100

# Keep only the relevant columns
season_df = season_df[['branded_code_share','season','year','month',
                       'Avg. min delayed for arrival flights (season)',
                       'Avg. min delayed for departure flights (season)',
                       'arrival flights season (%)','departure flights season (%)']]
season_df

branded_code_share  season  year  month  \
0                   AA    fall  2022      9   
1                   AA    fall  2022     10   
2                   AA    fall  2022     11   
3                   AA    fall  2023      9   
4                   AA    fall  2023     10   
..                 ...     ...   ...    ...   
319                 WN  winter  2023      2   
320                 WN  winter  2023     12   
321                 WN  winter  2024      1   
322                 WN  winter  2024      2   
323                 WN  winter  2024     12   

     Avg. min delayed for arrival flights (season)  \
0                                        83.731183   
1                                        91.716814   
2                                        86.949580   
3                                        86.029412   
4                                        79.929078   
..                                             ...   
319                                      47.360215   
320                                      56.094872   
321                                      65.337868   
322                                      52.410853   
323                                      54.943890   

     Avg. min delayed for departure flights (season)  \
0                                          78.465347   
1                                          88.778689   
2                                          99.171717   
3                                          94.666667   
4                                          88.361538   
..                                               ...   
319                                        48.867102   
320                                        54.431647   
321                                        60.901822   
322                                        50.600437   
323                                        54.433868   

     arrival flights season (%)  departure flights season (%)  
0                         14.31                         15.54  
1                         14.75                         15.93  
2                         16.76                         13.94  
3                         21.66                         19.11  
4                         17.98                         16.58  
..                          ...                           ...  
319                       10.87                         13.41  
320                       18.01                         21.45  
321                       22.73                         25.46  
322                       11.35                         13.43  
323                       18.65                         23.20  

[324 rows x 8 columns]

# Time of day DF

In [73]:
# time_df1: Average minutes delayed by time of day, with year and month
time_df1 = (bna_df.groupby(['branded_code_share', 'time_of_day', 'year', 'month'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df1['Avg. min delayed for arrival flights (time)'] = time_df1['arr_delay_new'] / time_df1['arr_del15']
time_df1['Avg. min delayed for departure flights (time)'] = time_df1['dep_delay_new'] / time_df1['dep_del15']
time_df1 = time_df1[['branded_code_share','time_of_day','year','month',
                     'Avg. min delayed for arrival flights (time)',
                     'Avg. min delayed for departure flights (time)']]

# time_df2: Percent of flights delayed by time of day, with year and month
time_df2 = (bna_df.groupby(['branded_code_share', 'time_of_day', 'year', 'month'])
            [['arr_del15','dep_del15','flights']]
            .sum()
            .reset_index())
time_df2['arrival flights time (%)'] = (time_df2['arr_del15'] / time_df2['flights']).round(4) * 100
time_df2['departure flights time (%)'] = (time_df2['dep_del15'] / time_df2['flights']).round(4) * 100
time_df2 = time_df2[['branded_code_share','time_of_day','year','month',
                     'arrival flights time (%)','departure flights time (%)']]

# time_df3: Yearly average minutes delayed
time_df3 = (bna_df.groupby(['branded_code_share','year'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df3['Avg. min delayed for arrival flights (year)'] = time_df3['arr_delay_new'] / time_df3['arr_del15']
time_df3['Avg. min delayed for departure flights (year)'] = time_df3['dep_delay_new'] / time_df3['dep_del15']
time_df3 = time_df3[['branded_code_share','year',
                     'Avg. min delayed for arrival flights (year)',
                     'Avg. min delayed for departure flights (year)']]

# time_df4: Monthly average minutes delayed
time_df4 = (bna_df.groupby(['branded_code_share','year','month'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df4['Avg. min delayed for arrival flights (month)'] = time_df4['arr_delay_new'] / time_df4['arr_del15']
time_df4['Avg. min delayed for departure flights (month)'] = time_df4['dep_delay_new'] / time_df4['dep_del15']
time_df4 = time_df4[['branded_code_share','year','month',
                     'Avg. min delayed for arrival flights (month)',
                     'Avg. min delayed for departure flights (month)']]

# Merge all metrics together
time_df = time_df1.merge(time_df2, on=['branded_code_share','time_of_day','year','month'], how='outer') \
                  .merge(time_df3, on=['branded_code_share','year'], how='outer') \
                  .merge(time_df4, on=['branded_code_share','year','month'], how='outer')

# Sort by year, month, and time_of_day
time_df = time_df.sort_values(by=['year','month','time_of_day'])
time_df

branded_code_share time_of_day  year  month  \
0                   AA       Night  2022      1   
36                  AS       Night  2022      1   
72                  B6       Night  2022      1   
108                 DL       Night  2022      1   
144                 F9       Night  2022      1   
..                 ...         ...   ...    ...   
179                 F9       Night  2024     12   
215                 G4       Night  2024     12   
251                 NK       Night  2024     12   
287                 UA       Night  2024     12   
323                 WN       Night  2024     12   

     Avg. min delayed for arrival flights (time)  \
0                                      95.009174   
36                                     73.833333   
72                                     87.750000   
108                                   120.589041   
144                                    61.545455   
..                                           ...   
179                                   120.333333   
215                                   107.523810   
251                                    72.410256   
287                                    60.916667   
323                                    54.943890   

     Avg. min delayed for departure flights (time)  arrival flights time (%)  \
0                                       101.000000                     17.67   
36                                       52.363636                     30.00   
72                                       86.000000                     33.33   
108                                     130.525424                     14.31   
144                                      53.454545                     22.45   
..                                             ...                       ...   
179                                     104.444444                     23.68   
215                                      99.177778                     20.29   
251                                      76.675000                     20.86   
287                                      65.500000                     18.67   
323                                      54.433868                     18.65   

     departure flights time (%)  Avg. min delayed for arrival flights (year)  \
0                         15.24                                    92.353870   
36                        27.50                                    55.078125   
72                        26.19                                    81.847411   
108                       11.57                                    86.090216   
144                       22.45                                    60.720848   
..                          ...                                          ...   
179                       23.68                                    89.744186   
215                       21.74                                   107.210526   
251                       21.39                                    82.326327   
287                       15.56                                    92.205989   
323                       23.20                                    58.228019   

     Avg. min delayed for departure flights (year)  \
0                                        90.602456   
36                                       54.786885   
72                                       80.790859   
108                                      85.958034   
144                                      62.731449   
..                                             ...   
179                                      93.270718   
215                                     106.288703   
251                                      82.137897   
287                                      96.556213   
323                                      56.676116   

     Avg. min delayed for arrival flights (month)  \
0                                       95.009174   
36                                      73.833333   
72                                  

# Day of the week DF

In [75]:
#day_df1: Average minutes delayed by day of the week, with year and month
day_df1 = (bna_df.groupby(['branded_code_share', 'day', 'year', 'month'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
day_df1['Avg. min delayed for arrival flights (day)'] = day_df1['arr_delay_new'] / day_df1['arr_del15']
day_df1['Avg. min delayed for departure flights (day)'] = day_df1['dep_delay_new'] / day_df1['dep_del15']
day_df1 = day_df1[['branded_code_share','day','year','month',
                   'Avg. min delayed for arrival flights (day)',
                   'Avg. min delayed for departure flights (day)']]

# day_df2: Percent of flights delayed by day of the week, with year and month
day_df2 = (bna_df.groupby(['branded_code_share','day','year','month'])
            [['arr_del15','dep_del15','flights']]
            .sum()
            .reset_index())
day_df2['arrival flights week (%)'] = (day_df2['arr_del15'] / day_df2['flights']).round(4) * 100
day_df2['departure flights week (%)'] = (day_df2['dep_del15'] / day_df2['flights']).round(4) * 100
day_df2 = day_df2[['branded_code_share','day','year','month',
                   'arrival flights week (%)','departure flights week (%)']]

# day_df3: Yearly average minutes delayed
day_df3 = (bna_df.groupby(['branded_code_share','year'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
day_df3['Avg. min delayed for arrival flights (year)'] = day_df3['arr_delay_new'] / day_df3['arr_del15']
day_df3['Avg. min delayed for departure flights (year)'] = day_df3['dep_delay_new'] / day_df3['dep_del15']
day_df3 = day_df3[['branded_code_share','year',
                   'Avg. min delayed for arrival flights (year)',
                   'Avg. min delayed for departure flights (year)']]

# Merge all metrics together
day_df = day_df1.merge(day_df2, on=['branded_code_share','day','year','month'], how='outer') \
                .merge(day_df3, on=['branded_code_share','year'], how='outer')

# Sort by year, month, and day
day_df = day_df.sort_values(by=['year','month','day'])
day_df

branded_code_share        day  year  month  \
0                    AA     Friday  2022      1   
252                  AS     Friday  2022      1   
504                  B6     Friday  2022      1   
756                  DL     Friday  2022      1   
1008                 F9     Friday  2022      1   
...                 ...        ...   ...    ...   
1007                 DL  Wednesday  2024     12   
1254                 F9  Wednesday  2024     12   
1744                 NK  Wednesday  2024     12   
1996                 UA  Wednesday  2024     12   
2248                 WN  Wednesday  2024     12   

      Avg. min delayed for arrival flights (day)  \
0                                      68.187500   
252                                          inf   
504                                    50.000000   
756                                   144.533333   
1008                                   71.000000   
...                                          ...   
1007                                   49.000000   
1254                                         NaN   
1744                                   23.000000   
1996                                   64.333333   
2248                                   46.647059   

      Avg. min delayed for departure flights (day)  arrival flights week (%)  \
0                                        70.266667                     19.05   
252                                      37.000000                      0.00   
504                                      41.000000                     25.00   
756                                     158.333333                     20.55   
1008                                     30.500000                     12.50   
...                                            ...                       ...   
1007                                     73.000000                     11.49   
1254                                           NaN                      0.00   
1744                                     28.500000                     27.78   
1996                                     72.400000                     21.43   
2248                                     49.043478                     13.05   

      departure flights week (%)  Avg. min delayed for arrival flights (year)  \
0                          17.86                                    92.353870   
252                        20.00                                    55.078125   
504                        25.00                                    81.847411   
756                        16.44                                    86.090216   
1008                       25.00                                    60.720848   
...                          ...                                          ...   
1007                        6.90                                    81.674897   
1254                        0.00                                    89.744186   
1744                       11.11                                    82.326327   
1996                       17.86                                    92.205989   
2248                       17.66                                    58.228019   

      Avg. min delayed for departure flights (year)  
0                                         90.602456  
252                                       54.786885  
504                                       80.790859  
756                                       85.958034  
1008                                      62.731449  
...                                             ...  
1007                                      85.403067  
1254                                      93.270718  
1744                                      82.137897  
1996                                      96.556213  
2248                                      56.676116  

[2249 rows x 10 columns]

# Season DF

In [78]:
season_df1 = (bna_df.groupby(['branded_code_share', 'season','year','month'])
              [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
              .sum()
              .reset_index())
season_df1['Avg. min delayed for arrival flights (season)'] = season_df1['arr_delay_new'] / season_df1['arr_del15']
season_df1['Avg. min delayed for departure flights (season)'] = season_df1['dep_delay_new'] / season_df1['dep_del15']
season_df1 = season_df1[['branded_code_share','season','year','month',
                         'Avg. min delayed for arrival flights (season)',
                         'Avg. min delayed for departure flights (season)']]

# season_df2 (percentages, by season, year, and month)
season_df2 = (bna_df.groupby(['branded_code_share','season','year','month'])
              [['arr_del15','dep_del15','flights']]
              .sum()
              .reset_index())
season_df2['arrival flights season (%)'] = (season_df2['arr_del15'] / season_df2['flights']).round(4) * 100
season_df2['departure flights season (%)'] = (season_df2['dep_del15'] / season_df2['flights']).round(4) * 100
season_df2 = season_df2[['branded_code_share','season','year','month',
                         'arrival flights season (%)','departure flights season (%)']]

# season_df3 (yearly avg minutes delayed, by year)
season_df3 = (bna_df.groupby(['branded_code_share','year'])
              [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
              .sum()
              .reset_index())
season_df3['Avg. min delayed for arrival flights (year)'] = season_df3['arr_delay_new'] / season_df3['arr_del15']
season_df3['Avg. min delayed for departure flights (year)'] = season_df3['dep_delay_new'] / season_df3['dep_del15']
season_df3 = season_df3[['branded_code_share','year',
                         'Avg. min delayed for arrival flights (year)',
                         'Avg. min delayed for departure flights (year)']]

# season_df4 (monthly avg minutes delayed, by month and year)
season_df4 = (bna_df.groupby(['branded_code_share','year','month'])
              [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
              .sum()
              .reset_index())
season_df4['Avg. min delayed for arrival flights (month)'] = season_df4['arr_delay_new'] / season_df4['arr_del15']
season_df4['Avg. min delayed for departure flights (month)'] = season_df4['dep_delay_new'] / season_df4['dep_del15']
season_df4 = season_df4[['branded_code_share','year','month',
                         'Avg. min delayed for arrival flights (month)',
                         'Avg. min delayed for departure flights (month)']]

# Merge all
season_df = season_df1.merge(season_df2, on=['branded_code_share','season','year','month'], how='outer') \
                      .merge(season_df3, on=['branded_code_share','year'], how='outer') \
                      .merge(season_df4, on=['branded_code_share','year','month'], how='outer')

season_df = season_df.sort_values(by = ['year','month'])
season_df

branded_code_share  season  year  month  \
9                   AA  Winter  2022      1   
45                  AS  Winter  2022      1   
81                  B6  Winter  2022      1   
117                 DL  Winter  2022      1   
153                 F9  Winter  2022      1   
..                 ...     ...   ...    ...   
179                 F9  Winter  2024     12   
215                 G4  Winter  2024     12   
251                 NK  Winter  2024     12   
287                 UA  Winter  2024     12   
323                 WN  Winter  2024     12   

     Avg. min delayed for arrival flights (season)  \
9                                        95.009174   
45                                       73.833333   
81                                       87.750000   
117                                     120.589041   
153                                      61.545455   
..                                             ...   
179                                     120.333333   
215                                     107.523810   
251                                      72.410256   
287                                      60.916667   
323                                      54.943890   

     Avg. min delayed for departure flights (season)  \
9                                         101.000000   
45                                         52.363636   
81                                         86.000000   
117                                       130.525424   
153                                        53.454545   
..                                               ...   
179                                       104.444444   
215                                        99.177778   
251                                        76.675000   
287                                        65.500000   
323                                        54.433868   

     arrival flights season (%)  departure flights season (%)  \
9                         17.67                         15.24   
45                        30.00                         27.50   
81                        33.33                         26.19   
117                       14.31                         11.57   
153                       22.45                         22.45   
..                          ...                           ...   
179                       23.68                         23.68   
215                       20.29                         21.74   
251                       20.86                         21.39   
287                       18.67                         15.56   
323                       18.65                         23.20   

     Avg. min delayed for arrival flights (year)  \
9                                      92.353870   
45                                     55.078125   
81                                     81.847411   
117                                    86.090216   
153                                    60.720848   
..                                           ...   
179                                    89.744186   
215                                   107.210526   
251                                    82.326327   
287                                    92.205989   
323                                    58.228019   

     Avg. min delayed for departure flights (year)  \
9                                        90.602456   
45                                       54.786885   
81                                       80.790859   
117                                      85.958034   
153                                      62.731449   
..                                             ...   
179                                      93.270718   
215                                     106.288703   
251                                      82.137897   
287                                      96.556213   
323                                      56.676116   

     Avg. min delayed for arrival flights (month)  \
9        

In [92]:
# time_df1 (avg minutes delayed, by time of day, month, and year)
time_df1 = (bna_df.groupby(['branded_code_share','time_of_day','year','month'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df1['Avg. min delayed for arrival flights (time)'] = time_df1['arr_delay_new'] / time_df1['arr_del15']
time_df1['Avg. min delayed for departure flights (time)'] = time_df1['dep_delay_new'] / time_df1['dep_del15']
time_df1 = time_df1[['branded_code_share','time_of_day','year','month',
                     'Avg. min delayed for arrival flights (time)',
                     'Avg. min delayed for departure flights (time)']]

# time_df2 (percentages, by time of day, month, and year)
time_df2 = (bna_df.groupby(['branded_code_share','time_of_day','year','month'])
            [['arr_del15','dep_del15','flights']]
            .sum()
            .reset_index())
time_df2['arrival flights time (%)'] = (time_df2['arr_del15'] / time_df2['flights']).round(4) * 100
time_df2['departure flights time (%)'] = (time_df2['dep_del15'] / time_df2['flights']).round(4) * 100
time_df2 = time_df2[['branded_code_share','time_of_day','year','month','arrival flights time (%)','departure flights time (%)']]

# time_df3 (yearly avg minutes delayed, by year)
time_df3 = (bna_df.groupby(['branded_code_share','year'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df3['Avg. min delayed for arrival flights (year)'] = time_df3['arr_delay_new'] / time_df3['arr_del15']
time_df3['Avg. min delayed for departure flights (year)'] = time_df3['dep_delay_new'] / time_df3['dep_del15']
time_df3 = time_df3[['branded_code_share','year','Avg. min delayed for arrival flights (year)','Avg. min delayed for departure flights (year)']]

# time_df4 (monthly avg minutes delayed, by month and year)
time_df4 = (bna_df.groupby(['branded_code_share','year','month'])
            [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
            .sum()
            .reset_index())
time_df4['Avg. min delayed for arrival flights (month)'] = time_df4['arr_delay_new'] / time_df4['arr_del15']
time_df4['Avg. min delayed for departure flights (month)'] = time_df4['dep_delay_new'] / time_df4['dep_del15']
time_df4 = time_df4[['branded_code_share','year','month',
                     'Avg. min delayed for arrival flights (month)',
                     'Avg. min delayed for departure flights (month)']]

# Merge all
time_df = time_df1.merge(time_df2, on=['branded_code_share','time_of_day','year','month'], how='outer') \
                  .merge(time_df3, on=['branded_code_share','year'], how='outer') \
                  .merge(time_df4, on=['branded_code_share','year','month'], how='outer')

time_df = time_df.sort_values(by=['year','month','time_of_day'])


In [93]:
time_df 

branded_code_share time_of_day  year  month  \
0                    AA   Afternoon  2022      1   
315                  B6   Afternoon  2022      1   
493                  DL   Afternoon  2022      1   
709                  F9   Afternoon  2022      1   
891                  G4   Afternoon  2022      1   
...                 ...         ...   ...    ...   
1794                 WN     Morning  2024     12   
484                  B6       Night  2024     12   
1335                 NK       Night  2024     12   
1543                 UA       Night  2024     12   
1795                 WN       Night  2024     12   

      Avg. min delayed for arrival flights (time)  \
0                                      116.500000   
315                                    137.300000   
493                                    193.600000   
709                                     45.000000   
891                                    134.166667   
...                                           ...   
1794                                    61.257576   
484                                     32.500000   
1335                                    93.000000   
1543                                          NaN   
1795                                    47.351351   

      Avg. min delayed for departure flights (time)  arrival flights time (%)  \
0                                         97.875000                     15.38   
315                                      134.625000                     37.04   
493                                      139.500000                     15.62   
709                                       34.250000                     26.67   
891                                      121.833333                     44.44   
...                                             ...                       ...   
1794                                      57.812500                      8.47   
484                                       43.500000                    100.00   
1335                                      83.666667                     23.53   
1543                                            inf                      0.00   
1795                                      43.400000                     40.22   

      departure flights time (%)  Avg. min delayed for arrival flights (year)  \
0                          17.58                                    92.353870   
315                        29.63                                    81.847411   
493                        18.75                                    86.090216   
709                        26.67                                    60.720848   
891                        44.44                                    70.125846   
...                          ...                                          ...   
1794                       10.27                                    58.228019   
484                       100.00                                    86.964401   
1335                       35.29                                    82.326327   
1543                        0.00                                    92.205989   
1795                       59.78                                    58.228019   

      Avg. min delayed for departure flights (year)  \
0                                         90.602456   
315                                       80.790859   
493                                       85.958034   
709                                       62.731449   
891                                       69.356495   
...                                             ...   
1794                                      56.676116   
484                                       89.682390   
1335                                      82.137897   
1543                                      96.556213   
1795                                      56.676116   

      Avg. min delayed for arrival flights (month)  \
0                                        95.009174   
315                         

# Weekend/Weekday DF

In [77]:
# week_type_df1: Average minutes delayed by week_type, with year and month
week_type_df1 = (bna_df.groupby(['branded_code_share', 'week_type', 'year', 'month'])
                 [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
                 .sum()
                 .reset_index())
week_type_df1['Avg. min delayed for arrival flights (week_type)'] = week_type_df1['arr_delay_new'] / week_type_df1['arr_del15']
week_type_df1['Avg. min delayed for departure flights (week_type)'] = week_type_df1['dep_delay_new'] / week_type_df1['dep_del15']
week_type_df1 = week_type_df1[['branded_code_share','week_type','year','month',
                               'Avg. min delayed for arrival flights (week_type)',
                               'Avg. min delayed for departure flights (week_type)']]

# week_type_df2: Percent of flights delayed by week_type
week_type_df2 = (bna_df.groupby(['branded_code_share','week_type','year','month'])
                 [['arr_del15','dep_del15','flights']]
                 .sum()
                 .reset_index())
week_type_df2['arrival flights week_type (%)'] = (week_type_df2['arr_del15'] / week_type_df2['flights']).round(4) * 100
week_type_df2['departure flights week_type (%)'] = (week_type_df2['dep_del15'] / week_type_df2['flights']).round(4) * 100
week_type_df2 = week_type_df2[['branded_code_share','week_type','year','month',
                               'arrival flights week_type (%)','departure flights week_type (%)']]

# week_type_df3: Yearly averages
week_type_df3 = (bna_df.groupby(['branded_code_share','year'])
                 [['arr_delay_new','dep_delay_new','arr_del15','dep_del15']]
                 .sum()
                 .reset_index())
week_type_df3['Avg. min delayed for arrival flights (year)'] = week_type_df3['arr_delay_new'] / week_type_df3['arr_del15']
week_type_df3['Avg. min delayed for departure flights (year)'] = week_type_df3['dep_delay_new'] / week_type_df3['dep_del15']
week_type_df3 = week_type_df3[['branded_code_share','year',
                               'Avg. min delayed for arrival flights (year)',
                               'Avg. min delayed for departure flights (year)']]

# Merge all metrics together
week_type_df = week_type_df1.merge(week_type_df2, on=['branded_code_share','week_type','year','month'], how='outer') \
                            .merge(week_type_df3, on=['branded_code_share','year'], how='outer')

# Sort by year, month, and week_type
week_type_df = week_type_df.sort_values(by=['year','month','week_type'])
week_type_df

branded_code_share week_type  year  month  \
0                   AA   weekday  2022      1   
72                  AS   weekday  2022      1   
144                 B6   weekday  2022      1   
216                 DL   weekday  2022      1   
288                 F9   weekday  2022      1   
..                 ...       ...   ...    ...   
359                 F9   weekend  2024     12   
431                 G4   weekend  2024     12   
503                 NK   weekend  2024     12   
575                 UA   weekend  2024     12   
647                 WN   weekend  2024     12   

     Avg. min delayed for arrival flights (week_type)  \
0                                           95.294118   
72                                          67.333333   
144                                         63.733333   
216                                        142.326923   
288                                         44.000000   
..                                                ...   
359                                        296.000000   
431                                         56.363636   
503                                         92.250000   
575                                         58.678571   
647                                         54.996599   

     Avg. min delayed for departure flights (week_type)  \
0                                           100.762712    
72                                           38.000000    
144                                          57.090909    
216                                         160.609756    
288                                          41.400000    
..                                                 ...    
359                                         173.500000    
431                                          51.384615    
503                                          96.153846    
575                                          60.043478    
647                                          55.307278    

     arrival flights week_type (%)  departure flights week_type (%)  \
0                            15.81                            13.72   
72                           21.43                            14.29   
144                          27.78                            20.37   
216                          14.21                            11.20   
288                          17.65                            14.71   
..                             ...                              ...   
359                           8.33                            16.67   
431                          18.03                            21.31   
503                          21.43                            23.21   
575                          21.71                            17.83   
647                          23.06                            29.10   

     Avg. min delayed for arrival flights (year)  \
0                                      92.353870   
72                                     55.078125   
144                                    81.847411   
216                                    86.090216   
288                                    60.720848   
..                                           ...   
359                                    89.744186   
431                                   107.210526   
503                                    82.326327   
575                                    92.205989   
647                                    58.228019   

     Avg. min delayed for departure flights (year)  
0                                        90.602456  
72                                       54.786885  
144                                      80.790859  
216                                      85.958034  
288                                      62.731449  
..                                             ...  
359                                      93.270718  
431                                     106.288703  
503                                      82.137897  
575          